pkg/engine/executor/executor.go

```diff
@@ -0,0 +1,287 @@
+			rows[i] = batch.NumRows()
+			col := batch.Column(2) // assuming timestamp column is at index 2
+			tsCol, ok := col.(*array.Uint64)
+			if !ok {
+				yield(nil, errors.New("column is not a timestamp column"))
+				return
+			}
+
+			for j := 0; int64(j) < batch.NumRows(); j++ {
+				row := Row{
+					value:   tsCol.Value(j),
+					rowIdx:  j,
+					iterIdx: i,
+				}
+				fmt.Printf("Push(row) %+v\n", row)
+				heap.Push(mh, row)
+			}
+		}
+
+		for mh.Len() > 0 {
+			row := heap.Pop(mh).(Row)
+			i := row.iterIdx
+			r := row.rowIdx
+
+			rows[i]--
+
+			fmt.Printf("yield(row) %+v\n", row)
+
+			// TODO(chaudum): This yields single-row batches!!!
+			if !yield(batches[row.iterIdx].NewSlice(int64(r), int64(r)+1), nil) {
+				return
+			}
+
+			if rows[i] <= 0 {
+				batches[i].Release()
+				next := nextFn[i]
+				batch, err, ok := next()
+				if !ok {
+					active[i] = false
+					fmt.Printf("no more results from downstream %d\n", i)
+					continue
+				}
+				if err != nil {
+					yield(nil, err)
+					return
+				}
+
+				batches[i] = batch
+				rows[i] = batch.NumRows()
+				fmt.Printf("new rows %d from input %d\n", rows[i], i)
+				col := batch.Column(2) // assuming timestamp column is at index 2
+				tsCol, ok := col.(*array.Uint64)
+				if !ok {
+					yield(nil, errors.New("column is not a timestamp column"))
+					return
+				}
+
+				for j := 0; int64(j) < batch.NumRows(); j++ {
+					row := Row{
+						value:   tsCol.Value(j),
+						rowIdx:  j,
+						iterIdx: i,
+					}
+					fmt.Printf("Push(row) %+v\n", row)
+					heap.Push(mh, row)
+				}
+			}
+		}
+
+	}
+}
+
+func (c *Context) executeLimit(ctx context.Context, n *physical.Limit, inputs []Pipeline) Pipeline {
+	if len(inputs) == 0 {
+		return emptyPipeline()
+	}
+	if len(inputs) > 1 {
+		return errorPipeline(fmt.Errorf("limit expects exactly one input, got %d", len(inputs)))
+	}
+
+	// We gradually reduce offset and limit as we process more records, as the
+	// offset and limit may cross record boundaries.
+	var (
+		offset = int64(n.Skip)
+		limit  = int64(n.Fetch)
+	)
+
+	input := inputs[0]
+	return func(yield func(arrow.Record, error) bool) {
+
+		// Stop once we reached the limit
+		if limit <= 0 {
+			return
+		}
+
+		// TODO(chaudum): Skip yielding zero-lenght batches while offset > 0
+
+		// Pull the next item from downstream
+		input(func(batch arrow.Record, err error) bool {
+			// If there's an error, pass it through
+			if err != nil {
+				return yield(nil, err)
+			}
+
+			// We want to slice batch so it only contains the rows we're looking for
+			// accounting for both the limit and offset.
+			// We constrain the start and end to be within the bounds of the record.
+			var (
+				start  = min(offset, batch.NumRows())
+				end    = min(start+limit, batch.NumRows())
+				length = end - start
+			)
+
+			offset -= start
+			limit -= length
+
+			if length <= 0 && offset <= 0 {
+				return false
+			}
+
+			return yield(batch.NewSlice(start, end), nil)
```

Comment:
NewSlice increases the reference count, so you need to release the record you have after. You can use a [CheckedAllocator](https://pkg.go.dev/github.com/apache/arrow-go/v18@v18.2.0/arrow/memory#CheckedAllocator) to make sure everything gets released properly after a pipeline executes